### ISSUE TITLE
Rule configuration is not working - architectural fix needed

### ISSUE THREAD

**Issue by joshrotenberg** [joshrotenberg/mdbook-lint#169]
# Rule Configuration is Not Working - Architectural Fix Needed

## Problem Description

Rule-specific configuration in mdbook-lint is completely non-functional. While configuration files are parsed and stored correctly, the configuration values are never actually applied to the rules. This affects all 21 configurable rules in the codebase.

### Current Behavior
- Users can specify rule configuration in `.mdbook-lint.toml` files
- Configuration is correctly parsed and stored in `Config.rule_configs`
- Rules are always created with default values, ignoring user configuration
- Tests only verify that configuration is parsed, not that it's actually used

### Expected Behavior
Rules should use the configuration values specified by users, e.g., MD013 should respect configured line length.

## Root Cause Analysis

The issue stems from the rule initialization architecture:

1. **Rules are created at provider registration time** in `StandardRuleProvider::register_rules()`:
   ```rust
   registry.register(Box::new(md013::MD013::default())); // Always uses default!
   ```

2. **Configuration is only available at runtime** when checking documents:
   ```rust
   pub fn check_document_with_config(&self, document: &Document, config: &Config)
   ```

3. **No mechanism exists to pass configuration to rules** - the `register_rules` method has no access to configuration:
   ```rust
   fn register_rules(&self, registry: &mut RuleRegistry); // No config parameter!
   ```

## Affected Rules (21 total)

All rules with `with_*` configuration methods that are never called:
- MD002, MD003, MD004, MD007, MD009, MD010, MD012, MD013
- MD024, MD025, MD026, MD029, MD030, MD035, MD036
- MD043, MD044, MD046, MD048, MD049, MD050, MD055

## Evidence

Testing confirms configuration is ignored:
```bash
# Create test file with 101-character line
echo "This is a line that is exactly 100 characters long..." > test.md

# Configure MD013 for 120 characters
echo "[MD013]
line-length = 120" > .mdbook-lint.toml

# Run linter - still warns at default 80 characters!
mdbook-lint lint --enable MD013 test.md
# Output: Line length is 101 characters, expected no more than 80
```

### Option 1: Pass Configuration to Providers (Recommended)

Modify the `RuleProvider` trait to accept configuration:

```rust
// Change the trait
pub trait RuleProvider: Send + Sync {
    fn register_rules(&self, registry: &mut RuleRegistry, config: &Config);
    // OR add a new method to maintain compatibility
    fn register_rules_with_config(&self, registry: &mut RuleRegistry, config: &Config) {
        self.register_rules(registry); // Default to old behavior
    }
}

// Update StandardRuleProvider implementation
impl RuleProvider for StandardRuleProvider {
    fn register_rules(&self, registry: &mut RuleRegistry, config: &Config) {
        // Create MD013 with configuration
        let md013 = if let Some(md013_config) = config.get_rule_config("MD013") {
            let line_length = md013_config.get("line-length")
                .and_then(|v| v.as_integer())
                .unwrap_or(80) as usize;
            MD013::with_line_length(line_length)
        } else {
            MD013::default()
        };
        registry.register(Box::new(md013));
        // ... repeat for other configurable rules
    }
}
```

### Option 2: Factory Pattern

Create rules on-demand with configuration:

```rust
pub trait RuleFactory: Send + Sync {
    fn create_rule(&self, config: Option<&toml::Value>) -> Box<dyn Rule>;
}

impl RuleRegistry {
    pub fn register_factory(&mut self, id: &str, factory: Box<dyn RuleFactory>);
    pub fn create_rule(&self, id: &str, config: &Config) -> Option<Box<dyn Rule>>;
}
```

### Option 3: Mutable Rules with Configure Method

Add a configuration method to the Rule trait:

```rust
pub trait Rule: Send + Sync {
    fn configure(&mut self, config: &toml::Value) -> Result<()> {
        Ok(()) // Default no-op
    }
}
```

## Implementation Tasks

For Option 1 (recommended):

1. **Update Core Trait** (`crates/mdbook-lint-core/src/engine.rs`)
   - Modify `RuleProvider::register_rules` to accept `config: Option<&Config>`
   - Update `PluginRegistry::create_engine` to pass configuration

2. **Update Provider Implementations** (`crates/mdbook-lint-rulesets/src/`)
   - Modify `StandardRuleProvider::register_rules` to use configuration
   - Modify `MdBookRuleProvider::register_rules` similarly
   - Parse configuration and create rules with appropriate `with_*` methods

3. **Update Rule Creation Logic**
   - For each configurable rule, extract configuration and call appropriate constructor
   - Handle missing/invalid configuration gracefully with defaults

4. **Update All Callers**
   - Update all places that call `register_rules` to pass configuration
   - Ensure configuration is available at engine creation time

5. **Add Integration Tests**
   - Test that MD013 actually uses configured line length
   - Test configuration for other rules
   - Test invalid configuration handling

6. **Update Documentation**
   - Document how to configure rules
   - Provide examples for each configurable rule

## Breaking Changes

Option 1 would be a breaking change to the `RuleProvider` trait. We could minimize this by:
- Adding a new method `register_rules_with_config` with a default implementation
- Deprecating the old `register_rules` method
- Migrating over time

## Testing Strategy

1. **Unit Tests**: Test each rule's configuration parsing
2. **Integration Tests**: Test end-to-end configuration flow
3. **CLI Tests**: Test that CLI respects configuration files
4. **Regression Tests**: Ensure default behavior unchanged when no config provided

## Impact

This fix would:
- Make all 21 configurable rules actually configurable
- Fix issues #161 and similar configuration complaints
- Complete the configuration feature that users expect to work
- Improve user experience significantly

## Priority

HIGH - This is a fundamental feature that users expect to work and affects many rules.

## References

- Issue #161: Rule configuration not working
- All rules with `with_*` methods that are never called
- Configuration parsing code that works but is never used

**Comment by joshrotenberg** [comment]
This issue has been **RESOLVED** ✅

The configuration system was completely overhauled in PR #170 and subsequent PRs (#176-179). Rules now properly receive and apply configuration settings.

## What was fixed:
- Added `register_rules_with_config` method to the `RuleProvider` trait
- Implemented `from_config` methods for all configurable rules
- Updated `PluginRegistry` to pass configuration during rule creation
- The engine now uses `create_engine_with_config` to properly initialize rules with user settings

## Verification:
Testing shows configuration now works correctly. For example, setting `line-length = 120` for MD013 properly overrides the default 80-character limit.

## Related PRs:
- #170 - Initial rule configuration support
- #176 - Batch 1 rules (MD002, MD003, MD007, MD010, MD012)
- #177 - Batch 2 rules (MD024-MD030)
- #178 - Batch 3 rules (MD035, MD036, MD043, MD044, MD046)
- #179 - Batch 4 rules (MD048-MD059)

All 21+ configurable rules now properly respect user configuration.
